src/Analyzers/MSTest.Analyzers/PreferAssertFailOverAlwaysFalseConditionsAnalyzer.cs

```diff
@@ -0,0 +1,107 @@
+            INamedTypeSymbol? assertSymbol = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingAssert);
+            if (assertSymbol is not null)
+            {
+                context.RegisterOperationAction(context => AnalyzeOperation(context, assertSymbol), OperationKind.Invocation);
+            }
+        });
+    }
+
+    private static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol assertSymbol)
+    {
+        var operation = (IInvocationOperation)context.Operation;
+        if (assertSymbol.Equals(operation.TargetMethod?.ContainingType, SymbolEqualityComparer.Default) &&
+            IsAlwaysFalse(operation))
+        {
+            context.ReportDiagnostic(operation.CreateDiagnostic(Rule, operation.TargetMethod.Name));
+        }
+    }
+
+    private static bool IsAlwaysFalse(IInvocationOperation operation)
+        => operation.TargetMethod.Name switch
+        {
+            "IsTrue" => GetConditionArgument(operation) is { Value.ConstantValue: { HasValue: true, Value: false } },
+            "IsFalse" => GetConditionArgument(operation) is { Value.ConstantValue: { HasValue: true, Value: true } },
+            "AreEqual" => GetEqualityStatus(operation, ExpectedParameterName) == EqualityStatus.NotEqual,
+            "AreNotEqual" => GetEqualityStatus(operation, NotExpectedParameterName) == EqualityStatus.Equal,
+            "IsNotNull" => GetValueArgument(operation) is { Value.ConstantValue: { HasValue: true, Value: null } },
+            _ => false,
+        };
+
+    private static IArgumentOperation? GetArgumentWithName(IInvocationOperation operation, string name)
+        => operation.Arguments.FirstOrDefault(arg => arg.Parameter?.Name == name);
+
+    private static IArgumentOperation? GetConditionArgument(IInvocationOperation operation)
+        => GetArgumentWithName(operation, ConditionParameterName);
+
+    private static IArgumentOperation? GetValueArgument(IInvocationOperation operation)
+        => GetArgumentWithName(operation, ValueParameterName);
+
+    private static EqualityStatus GetEqualityStatus(IInvocationOperation operation, string expectedOrNotExpectedParameterName)
+    {
+#pragma warning disable IDE0046 // Convert to conditional expression - more readable this way.
+        if (GetArgumentWithName(operation, expectedOrNotExpectedParameterName) is { } expectedOrNotExpectedArgument &&
+            GetArgumentWithName(operation, ActualParameterName) is { } actualArgument &&
+            expectedOrNotExpectedArgument.Value.ConstantValue.HasValue &&
+            actualArgument.Value.ConstantValue.HasValue)
+        {
+            return Equals(expectedOrNotExpectedArgument.Value.ConstantValue.Value, actualArgument.Value.ConstantValue.Value) ? EqualityStatus.Equal : EqualityStatus.NotEqual;
+        }
+#pragma warning restore IDE0046 // Convert to conditional expression
+
+        return EqualityStatus.Unknown;
```

Comment:
Fine as-is but just FYI you can use the "comment trick"

```suggestion
        if (GetArgumentWithName(operation, expectedOrNotExpectedParameterName) is { } expectedOrNotExpectedArgument &&
            GetArgumentWithName(operation, ActualParameterName) is { } actualArgument &&
            expectedOrNotExpectedArgument.Value.ConstantValue.HasValue &&
            actualArgument.Value.ConstantValue.HasValue)
        {
            return Equals(expectedOrNotExpectedArgument.Value.ConstantValue.Value, actualArgument.Value.ConstantValue.Value) ? EqualityStatus.Equal : EqualityStatus.NotEqual;
        }

        // We are not sure about the equality status
        return EqualityStatus.Unknown;
```